redis/src/cluster_async/mod.rs

```diff
@@ -789,34 +797,75 @@ where
+        let topology_join_results =
+            futures::future::join_all(requested_nodes.map(|node| async move {
+                let mut conn: C = node.1.await;
+                (node.0, conn.req_packed_command(&slot_cmd()).await)
+            }))
+            .await;
+        let mut topology_values: Vec<_> = vec![];
+        let mut failed_connections: Vec<_> = vec![];
+        for (identifier, result) in topology_join_results.into_iter() {
+            match result {
+                Ok(res) => topology_values.push(res),
+                Err(err) => {
+                    if err.is_connection_dropped() {
+                        failed_connections.push(identifier);
+                    }
+                }
+            }
         }
+        (topology_values, failed_connections)
     }
 
     /// Queries log2n nodes (where n represents the number of cluster nodes) to determine whether their
     /// topology view differs from the one currently stored in the connection manager.
-    /// Returns true if change was detected, otherwise false.
-    async fn check_for_topology_diff(inner: Arc<InnerCore<C>>) -> RedisResult<bool> {
+    /// returns a tuple containing two elements: a boolean indicating whether a change was detected, and a vector containing identifiers of failed connections.
+    async fn check_for_topology_diff(
+        inner: Arc<InnerCore<C>>,
+    ) -> RedisResult<(bool, Vec<ConnectionIdentifier>)> {
         let read_guard = inner.conn_lock.read().await;
-        let num_of_nodes: usize = read_guard.len();
         // TODO: Starting from Rust V1.67, integers has logarithms support.
         // When we no longer need to support Rust versions < 1.67, remove fast_math and transition to the ilog2 function.
         let num_of_nodes_to_query =
-            std::cmp::max(fast_math::log2_raw(num_of_nodes as f32) as usize, 1);
-        let requested_nodes =
-            read_guard.random_connections(num_of_nodes_to_query, ConnectionType::User);
-        let topology_join_results =
-            futures::future::join_all(requested_nodes.map(|conn| async move {
-                let mut conn: C = conn.1.await;
-                conn.req_packed_command(&slot_cmd()).await
-            }))
-            .await;
-        let topology_values: Vec<_> = topology_join_results
-            .into_iter()
-            .filter_map(|r| r.ok())
-            .collect();
+            std::cmp::max(fast_math::log2_raw(read_guard.len() as f32) as usize, 1);
+        let (topology_values, failed_connections) = Self::get_topology_values_and_failed_conn(
+            &read_guard,
+            num_of_nodes_to_query,
+            ConnectionType::Management,
+        )
+        .await;
+        if topology_values.is_empty() && !failed_connections.is_empty() {
+            // Couldn't get any topology views and failed connections were found.
+            // Return to the caller to refresh failed connections
+            return Ok((false, failed_connections));
```

Comment:
minor: consider returning an enum instead of a boolean - again, for clarity of intention.

Reply:
TopologyState::Changed / Unchanged or TopologyRefresh::Required / Unrequired, for example